cdap-app-fabric/src/main/java/io/cdap/cdap/internal/provision/DefaultProvisionerConfigProvider.java

```diff
@@ -47,55 +47,73 @@
-          DirUtils.listFiles(moduleDir,
-                             file -> {
-                               String name = file.getName();
-                               return !file.isDirectory() && name.endsWith(".json")
-                                 && provisioners.contains(name.substring(0, name.lastIndexOf('.')));
-                             });
-        if (jsonFiles.isEmpty()) {
-          LOG.info("Not able to find JSON config file for module {}", moduleDir);
+      try (Reader reader = openConfigReader(provisioner)) {
+        if (reader == null) {
           continue;
         }
-
-        for (File configFile : jsonFiles) {
-          String name = configFile.getName();
-          String provisionerName = name.substring(0, name.lastIndexOf('.'));
-          try (Reader reader = Files.newReader(configFile, Charsets.UTF_8)) {
-            results.put(provisionerName, GSON.fromJson(new JsonReader(reader), ProvisionerConfig.class));
-          } catch (Exception e) {
-            LOG.warn("Exception reading JSON config file for provisioner {}. Ignoring file.",
-                     provisionerName, e);
-          }
-        }
+        results.put(provisionerName, GSON.fromJson(new JsonReader(reader), ProvisionerConfig.class));
+      } catch (Exception e) {
+        LOG.warn("Exception reading JSON config file for provisioner {}. Ignoring file.", provisionerName, e);
       }
     }
+
     return Collections.unmodifiableMap(results);
   }
+
+  /**
+   * Find the json config file with name [provisioner_name].json under the provisioner ext directory.
+   * The ext directory can be found by the classloader of the provisioner.
+   * If the json config file is not found in the ext directory, try to locate it from the provisioner classloader.
+   *
+   * @param provisioner the {@link Provisioner}
+   * @return a {@link Reader} for reading the config file, or {@code null} if the config file cannot be located
+   */
+  @Nullable
+  private Reader openConfigReader(Provisioner provisioner) throws IOException, URISyntaxException {
+    Class<? extends Provisioner> provisionerClass = provisioner.getClass();
+    ClassLoader classLoader = provisionerClass.getClassLoader();
+    URL classURL = classLoader.getResource(provisionerClass.getName().replace('.', '/') + ".class");
+
+    // This shouldn't happen as we should be able to find the class from its own classloader
+    if (classURL == null) {
+      LOG.warn("Unable to find provisioner class {} from its own classloader", provisionerClass);
+      return null;
+    }
+
+    // Find the json config file with name [provisioner_name].json under the provisioner ext directory.
+    // The ext directory can be found by the classloader of the provisioner.
+    // See AbstractExtensionLoader class for the directory structure
+    String provisionerName = provisioner.getSpec().getName();
+    URL classPathURL = ClassLoaders.getClassPathURL(provisionerClass.getName(), classURL);
```

Comment:
Can you just use `ClassLoaders.getClassPathURL(Class<?> cls)` as it does same thing as two-step above?